Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/services/DirectoryInitializationService.java

```diff
@@ -55,21 +61,51 @@ public static void startService(Context context)
     @Override
     protected void onHandleIntent(Intent intent)
     {
+        isDolphinDirectoryInitializationRunning.set(true);
+
         if (directoryState == DirectoryInitializationState.DOLPHIN_DIRECTORIES_INITIALIZED)
         {
             sendBroadcastState(DirectoryInitializationState.DOLPHIN_DIRECTORIES_INITIALIZED);
         }
         else if (PermissionsHandler.hasWriteAccess(this))
         {
-            initializeInternalStorage();
-            initializeExternalStorage();
-            directoryState = DirectoryInitializationState.DOLPHIN_DIRECTORIES_INITIALIZED;
-            sendBroadcastState(directoryState);
+            if (setDolphinUserDirectory())
+            {
+                initializeInternalStorage();
+                initializeExternalStorage();
+                directoryState = DirectoryInitializationState.DOLPHIN_DIRECTORIES_INITIALIZED;
+                sendBroadcastState(directoryState);
+            }
+            else
+            {
+                directoryState = DirectoryInitializationState.CANT_FIND_EXTERNAL_STORAGE;
+                sendBroadcastState(directoryState);
+            }
         }
         else
         {
             sendBroadcastState(DirectoryInitializationState.EXTERNAL_STORAGE_PERMISSION_NEEDED);
         }
+
+        isDolphinDirectoryInitializationRunning.set(false);
```

Comment:
Isn't it a race condition that we set this after calling `sendBroadcastState` rather than before?

Reply:
You are right, I will have to call this under each block and it will look ugly. I don't want to add any synchronization to the whole method.

Reply:
Instead of moving this line into each block, you should be able to run `sendBroadcastState(directoryState)` at the very end of the method.